sympy/vector/dyadic.py

```diff
@@ -225,6 +225,13 @@ def _sympystr(self, printer):
         return "({}|{})".format(
             printer._print(self.args[0]), printer._print(self.args[1]))
 
+    #this should not be needed, but this class is indicated as an atom
+    #eventhough, it has `Basic` args, this makes the argument invariance
+    #behave as if it is an atom. It might be better to redesign this class
+    #since it uses conflicting designs.
+    @property
+    def func(self):
+        return lambda *args, **kwargs: self
```

Comment:
Why is this needed for `BaseDyadic`? The class seems to work okay to me:
```python
In [11]: from sympy.vector import CoordSys3D

In [12]: N = CoordSys3D('N')

In [13]: d = N.i | N.j

In [14]: type(d)
Out[14]: sympy.vector.dyadic.BaseDyadic

In [15]: d.func
Out[15]: sympy.vector.dyadic.BaseDyadic

In [16]: d.args
Out[16]: (i_N, j_N)

In [17]: d.func(*d.args)
Out[17]: (i_N|j_N)

In [18]: d.func(*d.args) == d
Out[18]: True
```

Reply:
`isinstance(d, Atom)` is `True` since it is an `AtomicExpr`. This means it will no longer work when `Atom` is updated to be argument invariant (if done so according to #22310). One can ask the question why this is an `AtomicExpr` in the first place, since it doesn't behave atomic at all)

Reply:
Maybe just remove the `AtomicExpr` subclass. The other superclass already subclasses `Expr` whch looks like the correct superclass to me.